File: Python/LinkedList/SingleLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SingleLinkedList:
    def __init__(self, data):
        self.head = Node(data)
        self.length = 1
        
    def checkEmpty(self) -> bool: #helper function to check if list is empty
        return self.head is None 
    
    def printList(self) -> None: #function to print the list
        if self.checkEmpty():
            print("THE LIST IS EMPTY")
        else:
            temp = self.head
            while temp != None:
                print(temp.data)
                temp = temp.next
            print("END OF LIST")

    def addFirst(self, data) -> None: #function to add Node at the stat of the list
        newNode = Node(data)
        if self.checkEmpty():
            self.head = newNode
        else:
            # set newNode 'next' to head
            # update 'head' to newNode
            newNode.next = self.head
            self.head = newNode
            
            self.length += 1

    def addLast(self, data) -> None: #function to add Node at the end of the list
        newNode = Node(data)
        if self.checkEmpty():
            self.head = newNode
        else:
            temp = self.head
            # traverse till last node
            while temp.next != None:
                temp = temp.next
            # set last node 'next' as newNode
            temp.next = newNode
            
            self.length += 1

    def deleteFirst(self) -> None: #function to delete Node at the start of the list
        if self.checkEmpty():
            print("THE LIST IS EMPTY")
        else:
            temp = self.head
            # update head to next node
            self.head = temp.next
            # delete starting node
            del temp

            self.length -= 1

    def deleteLast(self) -> None: #function to delete Node at the end of the list
        if self.checkEmpty():
            print("THE LIST IS EMPTY")
        else:
            temp = self.head
            # shadow to keep track of previous node
            shadow = None
            while temp.next != None:
                # shadow holds node before update
                shadow = temp
                # update to next node
                temp = temp.next
            # set second last node 'next' as None
            shadow.next = None
            # delete last node
            del temp

            self.length -= 1

    def join(self, list) -> None: #function to join two lists
        if self.checkEmpty() or list.checkEmpty():
            print("THE LIST IS EMPTY")
        else:
            temp = self.head
            while temp.next != None:
                temp = temp.next
            # set last node 'next' to list to be joined 'head'
            temp.next = list.head
            # update length of list
            self.length += list.length

    def searchList(self, target) -> bool: # method to return if 'target' is present in the list
        if self.checkEmpty():
            print("LIST IS EMPTY")
        else:
            temp = self.head
            while temp != None:
                if temp.data == target:
                    return True
                temp = temp.next
            return False
    
    def get(self, index) -> int: # method to return data at 'index' in list
        if self.checkEmpty():
            print("LIST IS EMPTY")
        else:
            if index >= self.length:
                return None
            else:
                temp = self.head
                num = 0
                while temp != None:
                    if num == index:
                        return temp.data
                    num += 1
                    temp = temp.next
                return None

    def indexOf(self, target) -> int: # method to return index of 'target'
        if self.checkEmpty():
            print("LIST IS EMPTY")
        else:
            temp = self.head
            index = 0
            while temp != None:
                if temp.data == target:
                    return index
                index += 1
                temp = temp.next
            return None
                
    def clearList(self) -> None:
        if self.checkEmpty():
            print("LIST IS ALREADY CLEAR")
        else:
            self.head = None
            self.length = 0

File: C++/LinkedList/SingleLinkedList.cpp
#include<iostream>

class Node {
public:
    int data;
    Node* next;

    Node(int data) {
        this->data = data;
        this->next = NULL;
    }
};

class SingleLinkedList {
private:
    Node* head;

public:
    int length;

    SingleLinkedList() {
        head = NULL;
        length = 0;
    }

    SingleLinkedList(int data) {
        head = new Node(data);
        length++;
        head->next = NULL;
    }

    void addFirst(int data) {
        Node* newNode = new Node(data);
        if(head == NULL) {
            head = newNode;
        } else {
            newNode->next = head;
            head = newNode;
        }
        length++;
    }

    void addLast(int data) {
        Node* newNode = new Node(data);
        if(head ==  NULL) {
            head = newNode;
        } else {
            Node* temp = head;
            while(temp->next != NULL)
                temp = temp->next;
            temp->next = newNode;
        }
        length++;
    }

    void printList() {
        if(head == NULL) {
            std::cout << "LIST IS EMPTY" << std::endl;
        } else {
            Node* temp = head;
            while(temp != NULL) {
                std::cout << temp->data << '\n';
                temp = temp->next;
            }
            std::cout << "\nEND OF LIST\n";
        }
    }

    void deleteFirst() {
        if(head == NULL) 
            std::cout << "LIST IS EMPTY\n";
        else {
            Node* temp = head;
            head = temp->next;
            delete temp;
            length--;
        }
    }

    void deleteLast() {
        if(head == NULL)
            std::cout << "LIST IS EMPTY\n";
        else if(head->next == NULL) {
            delete head;
            head = NULL;
            length--;
        } else {
            Node* temp = head;
            Node* shadow = NULL;
            while(temp->next != NULL) {
                shadow = temp;
                temp = temp->next;
            }
            shadow->next = NULL;
            delete temp;
            length--;
        }
    }

    void addMiddle(int data, int value) {
        Node* newNode = new Node(data);
        if(head == NULL) {
            std::cout << "EMPTY LIST\n";
            return;
        }
        else {
            Node* temp = head;
            Node* shadow = NULL;
            while(temp != NULL) {
                shadow = temp;
                temp = temp->next;
                if(shadow->data == value) {
                    shadow->next = newNode;
                    newNode->next = temp;
                    length++;
                    return;
                }
            }
            std::cout << "\nINSERTION FAILED, could not find value\n";
        }
    }

    void deleteMiddle(int value) {
        if(head == NULL)
            std::cout << "EMPTY LIST\n";
        else {
            Node* temp = head;
            Node* shadow = NULL;
            while(temp != NULL) {
                shadow = temp;
                temp = temp->next;
                if(shadow->data == value) {
                    shadow->next = temp->next;
                    delete temp;
                    length--;
                    return;
                }
            }
            std::cout << "\nDELETION FAILED, could not find value\n";
        }
    } 

    int get(int index) {
        if(head == NULL) {
            std::cout << "EMPTY LIST\n";
            return INT_MIN;
        }
        else {
            if(index >= length)
                return INT_MIN;
            Node* temp = head;
            while(index > 0) {
                temp = temp->next;
                index--;
            }
            return temp->data;
        }
    }

    int indexOf(int target) {
        if(head == NULL) {
            std::cout << "EMPTY LIST\n";
            return -1;
        } else {
            Node* temp = head;
            int index {0};
            while(temp != NULL) {
                if(temp->data == target) 
                    return index;
                temp = temp->next;
                index++;
            }
            return -1;
        }
    }

    void join(SingleLinkedList *list) {
        if(head == NULL || list->head == NULL) 
            std::cout << "EMPTY LIST\n";
        else {
            Node* temp = head;
            while(temp->next != NULL)
                temp = temp->next;
            temp->next = list->head;
        }
    }

    void clearList() {
        if(head ==  NULL) 
            std::cout << "LIST IS ALREADY CLEAR\n";
        else {
            Node* temp = head;
            Node* shadow = NULL;
            while(temp != NULL) {
                shadow = temp;
                temp = temp->next;
                delete shadow;
            }
            head = NULL;
        }
    }

    void reverseList() {
        if(head == NULL) 
            std::cout << "EMPTY LIST\n";
        else {
            SingleLinkedList* rev = new SingleLinkedList();
            Node* temp = head;
            while(temp != NULL) {
                rev->addFirst(temp->data);
                temp = temp->next;
            }
            this->clearList();
            head = rev->head;
        }
    }

    ~SingleLinkedList() {
        clearList();
    }
};

/*

    ADD main() 
    OR 
    INCLUDE TO USE SingleLinkedList

*/